File: frontend/src/types/schemas.ts
/**
 * frontend/src/types/schemas.ts
 * TypeScript mirror of INTERFACES.md §3 Pydantic schemas.
 * Keep in sync with backend/app/models/schemas.py.
 */

export type PartyRole = 'plaintiff' | 'defendant' | 'witness' | 'other'
export type ClaimStatus = 'SUPPORTED' | 'CONTRADICTED' | 'PARTIALLY_SUPPORTED' | 'UNVERIFIED'
export type SourceType = 'USER_PROVIDED' | 'PUBLIC_LEGAL_SOURCE' | 'WEB_SOURCE' | 'SYNTHETIC'
export type ProvenanceType = 'USER_PROVIDED' | 'PUBLIC_LEGAL_SOURCE' | 'SYNTHETIC'
export type JudgeProfileName = 'strict' | 'balanced' | 'skeptical'
export type TrialStatus = 'pending' | 'running' | 'paused' | 'completed' | 'error'

export interface Party {
  party_id: string
  name: string
  role: PartyRole
  description?: string
}

export interface CaseEvent {
  event_id: string
  description: string
  date?: string
  evidence_ids: string[]
}

export interface Claim {
  claim_id: string
  statement: string
  made_by: 'prosecution' | 'defense' | 'intake'
  related_evidence_ids: string[]
  status: ClaimStatus
}

export interface StructuredCase {
  case_id: string
  title: string
  description: string
  parties: Party[]
  claims: Claim[]
  events: CaseEvent[]
  legal_questions: string[]
  evidence_ids: string[]
  unknowns: string[]
  contradictions: string[]
  provenance_type: ProvenanceType
}

export interface EvidenceResult {
  evidence_id: string
  content: string
  source_type: SourceType
  document_id?: string
  document_page?: number
  relevance_score: number
}

export interface Argument {
  claim_id: string
  argument: string
  evidence_ids: string[]
  source_ids: string[]
  confidence: number
  side: 'prosecution' | 'defense'
  round: number
  responds_to_argument_id?: string
}

export interface FactCheck {
  claim_id: string
  status: ClaimStatus
  supporting_evidence_ids: string[]
  contradicting_evidence_ids: string[]
  confidence: number
  reasoning: string
}

export interface EvidenceQualityScore {
  evidence_id: string
  reliability: number
  directness: number
  relevance: number
  corroboration: number
  recency: number
  authenticity_notes?: string
  composite_score: number
  methodology_version: string
}

export interface CrossExaminationRound {
  round: number
  challenger: 'cross_examiner'
  target_argument_id: string
  question: string
  response: string
  outcome: 'strengthened' | 'weakened' | 'unchanged'
}

export interface JudgeProfile {
  name: JudgeProfileName
}

export interface Verdict {
  finding: string
  supporting_evidence_ids: string[]
  opposing_evidence_ids: string[]
  unresolved_questions: string[]
  reasoning: string
  confidence: number
  judge_profile: string
  disclaimer: string
}

export interface HumanIntervention {
  new_document_ids: string[]
  affected_claim_ids: string[]
  submitted_at: string
}

// API-layer types ─────────────────────────────────────────────────────────────

export interface TokenResponse {
  access_token: string
  refresh_token: string
  token_type: string
}

export interface CaseRow {
  case_id: string
  title: string
  description: string
  provenance_type: ProvenanceType
  owner_id: string
  created_at: string
  status: string
}

export interface DocumentRow {
  document_id: string
  case_id: string
  filename: string
  content_type: string
  size_bytes: number
  upload_status: string
  created_at: string
}

export interface TrialStateResponse {
  case_id: string
  status: TrialStatus
  round: number
  verdict?: Verdict
  state_snapshot?: Record<string, unknown>
}

// Evidence graph ──────────────────────────────────────────────────────────────

export type NodeType = 'claim' | 'evidence' | 'source' | 'document'

export interface GraphNode {
  id: string
  type: NodeType
  label: string
  data: Record<string, unknown>
}

export interface GraphEdge {
  id: string
  source: string
  target: string
  label?: string
}

export interface EvidenceGraphResponse {
  case_id: string
  nodes: GraphNode[]
  edges: GraphEdge[]
}

// Replay ──────────────────────────────────────────────────────────────────────

export interface AgentMessage {
  message_id: string
  case_id: string
  agent_name: string
  event_type: string
  content: string
  evidence_refs: string[]
  confidence?: number
  timestamp: string
}

export interface ReplayResponse {
  case_id: string
  events: AgentMessage[]
}

File: backend/app/api/schemas.py
"""HTTP-only models owned by the API layer.

The shared domain models remain exclusively in :mod:`backend.app.models.schemas`, as
defined by INTERFACES.md §3.  Keeping transport shapes here prevents the API
from depending on the stale nested ``app/app`` package that was introduced by
the merge.
"""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field

from backend.app.models.schemas import HumanIntervention, Verdict


class ErrorResponse(BaseModel):
    error_code: str
    message: str
    details: Optional[dict] = None


class RegisterRequest(BaseModel):
    email: str
    password: str
    full_name: Optional[str] = None


class LoginRequest(BaseModel):
    email: str
    password: str


class TokenResponse(BaseModel):
    access_token: str
    refresh_token: str
    token_type: str = "bearer"


class RefreshRequest(BaseModel):
    refresh_token: str


class CreateCaseRequest(BaseModel):
    title: str
    description: str
    provenance_type: Literal[
        "USER_PROVIDED", "PUBLIC_LEGAL_SOURCE", "SYNTHETIC"
    ] = "USER_PROVIDED"


class CaseRow(BaseModel):
    case_id: str
    title: str
    description: str
    provenance_type: str
    owner_id: str
    created_at: str
    status: str = "pending"


class StartTrialRequest(BaseModel):
    case_id: str
    judge_profile: Literal["strict", "balanced", "skeptical"] = "balanced"


class TrialStateResponse(BaseModel):
    case_id: str
    status: Literal["pending", "running", "paused", "completed", "error"]
    round: int = 0
    verdict: Optional[Verdict] = None
    state_snapshot: Optional[dict] = None


class InterventionRequest(BaseModel):
    case_id: str
    intervention: HumanIntervention


class PublicSearchRequest(BaseModel):
    query: str
    jurisdiction: Optional[str] = None
    date_from: Optional[str] = None
    date_to: Optional[str] = None
    top_k: int = 5


class GraphNode(BaseModel):
    id: str
    type: Literal["claim", "evidence", "source", "document"]
    label: str
    data: dict = Field(default_factory=dict)


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    label: Optional[str] = None


class EvidenceGraphResponse(BaseModel):
    case_id: str
    nodes: list[GraphNode]
    edges: list[GraphEdge]


class AgentMessage(BaseModel):
    message_id: str
    case_id: str
    agent_name: str
    event_type: str
    content: str
    evidence_refs: list[str] = Field(default_factory=list)
    confidence: Optional[float] = None
    timestamp: str


class ReplayResponse(BaseModel):
    case_id: str
    events: list[AgentMessage]
